### PR TITLE
Configuring optimization hyper-params in Perturber

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

This PR enables configuring perturbation-wise hyper-params in attack optimization, such as learning rate, momentum, etc.

The main changes include
* Add support to `**optim_params` in `Perturber`.
* `Adversary` can fetch parameters and their optimization hyper-params by calling `Perturber.parameter_groups()`.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by dxoigmn** [comment]
Is this not already supported via per-parameter options? See: https://pytorch.org/docs/stable/optim.html#per-parameter-options

**Comment by mzweilin** [comment]
> Is this not already supported via per-parameter options? See: https://pytorch.org/docs/stable/optim.html#per-parameter-options

Yes, it's supported. And this PR is trying to use this feature by constructing trainable parameters with optimization information.

File: mart/attack/adversary.py [hunk 1]
```diff
@@ -159,7 +159,13 @@ def on_run_start(
         self.perturber(input, target)
 
         # Split param groups by input elements, so that we can schedule optimizers individually.
-        param_groups = [{"params": [param]} for param in self.perturber.parameters()]
+        if hasattr(self.perturber, "parameters_optim"):
+            # param_groups with learning rate and other optim params.
+            param_groups = self.perturber.parameters_optim()
+        else:
+            # Backward compatibility.
+            param_groups = [{"params": [param]} for param in self.perturber.parameters()]
```
Comment (dxoigmn) [inline mart/attack/adversary.py:167]: Why is this necessary of all perturbers have changed to the new interface?

File: mart/attack/adversary.py [hunk 2]
```diff
@@ -159,7 +159,13 @@ def on_run_start(
         self.perturber(input, target)
 
         # Split param groups by input elements, so that we can schedule optimizers individually.
-        param_groups = [{"params": [param]} for param in self.perturber.parameters()]
+        if hasattr(self.perturber, "parameters_optim"):
+            # param_groups with learning rate and other optim params.
+            param_groups = self.perturber.parameters_optim()
```
Comment (dxoigmn) [inline mart/attack/adversary.py:164]: Maybe rename `parameter_optim` to `parameter_groups`?

File: mart/attack/perturber/perturber.py [hunk 3]
```diff
@@ -56,6 +59,13 @@ def projector_wrapper(perturber_module, args):
         if projector is not None:
             self.register_forward_pre_hook(projector_wrapper)
 
+    def parameters_optim(self):
+        """Return parameters along with the pre-defined optimization parameters.
+
+        Example: `{"params": perturbation, "lr":0.1, "momentum": 0.9}`
+        """
+        return [{"params": self.perturbation} | self.optim_params]
```
Comment (dxoigmn) [inline mart/attack/perturber/perturber.py:67]: You might want to make sure that `self.optim_params` does not contain a `params` key.

File: tests/test_adversary.py [hunk 4]
```diff
@@ -69,7 +69,6 @@ def test_adversary_with_model(input_data, target_data, perturbation):
 
     output_data = adversary(input_data, target_data, model=model)
 
-    parameters.assert_called_once()
```
Comment (dxoigmn) [inline tests/test_adversary.py:72]: Thou shalt not change tests. Create a new test?

File: mart/attack/adversary.py (hunk 1 above)
Comment (mzweilin) [reply mart/attack/adversary.py]: Deleted.

File: mart/attack/perturber/perturber.py (hunk 3 above)
Comment (mzweilin) [reply mart/attack/perturber/perturber.py:72]: Added.

File: tests/test_adversary.py (hunk 4 above)
Comment (mzweilin) [reply tests/test_adversary.py:72]: The test case is tailored for the previous API and we have to change it, otherwise it fails.

File: mart/attack/perturber/batch.py [hunk 5]
```diff
@@ -43,6 +43,13 @@ def __init__(
 
         self.perturbers = torch.nn.ModuleDict()
 
+    def parameter_groups(self):
+        """Return parameters along with optim parameters."""
+        params = []
+        for _, perturber in self.perturbers.items():
```
Comment (dxoigmn) [inline mart/attack/perturber/batch.py:49]: Why not just `self.perturbers.values()`?

**Review by dxoigmn — APPROVED** [review]
LGTM, although I would consider cleaning up the code mentioned in the comments.

File: mart/attack/perturber/batch.py (hunk 5 above)
Comment (mzweilin) [reply mart/attack/perturber/batch.py]: Replaced.
